### ISSUE TITLE
Validation of unions containing merged objects failing

### ISSUE THREAD

**Issue by flenter** [vriad/zod#112]
First of all: i really like this package (especially because of the typescript type integration)

However, I ran into an issue where i was validating a union of objects, which were based on the same `base` object. This object had some properties defined and the validation of those properties doesn't work correctly.

Given the following code:
``` ts
const base = z.object({
  id: z.string(),
});

const type1 = base.merge(
  z.object({
    type: z.literal("type1"),
  })
);
const type2 = base.merge(
  z.object({
    type: z.literal("type2"),
  })
);

const union1 = z.union([type1, type2]);
const union2 = z.union([type2, type1]);

const value1 = {
  type: "type1",
};

const value2 = {
  type: "type2",
};

console.log("value1", value1);
console.table([
  ["type1 directly", type1.check(value1)],
  ["z.union([type1, type2])", union1.check(value1)],
  ["z.union([type2, type1])", union2.check(value1)],
]);

console.log("value2", value2);
console.table([
  ["type2 directly", type2.check(value2)],
  ["z.union([type1, type2])", union1.check(value2)],
  ["z.union([type2, type1])", union2.check(value2)],
]);
```
In the example objects above, the `id` props is left out completely

You get the following output:
``` sh
  console.log
    value1 { type: 'type1' }

  console.log
    ┌─────────┬───────────────────────────┬───────┐
    │ (index) │             0             │   1   │
    ├─────────┼───────────────────────────┼───────┤
    │    0    │     'type1 directly'      │ false │
    │    1    │ 'z.union([type1, type2])' │ false │
    │    2    │ 'z.union([type2, type1])' │ true  │
    └─────────┴───────────────────────────┴───────┘

  console.log
    value2 { type: 'type2' }

  console.log
    ┌─────────┬───────────────────────────┬───────┐
    │ (index) │             0             │   1   │
    ├─────────┼───────────────────────────┼───────┤
    │    0    │     'type2 directly'      │ false │
    │    1    │ 'z.union([type1, type2])' │ true  │
    │    2    │ 'z.union([type2, type1])' │ false │
    └─────────┴───────────────────────────┴───────┘
```

The first line in the table shows the expected result (false) but as you can see the unions aren't consistent in their result.

If i pass in an object where type is set to something that doesn't match any of the allowed values then validation works correctly. This means that check will return `false` no matter which order the `z.objects` are passed to `z.union`

**Comment by flenter** [comment]
I can fix the problem if i remove the `params` argument from the `options.parse` call in `src/parser.ts`

Change the last line of the code below
``` ts
    case z.ZodTypes.union:
      let isValid = false;
      const unionErrors: ZodError[] = [];
      for (const option of def.options) {
        try {
          option.parse(obj, params);
```

And remove the params argument like so:
``` ts
    case z.ZodTypes.union:
      let isValid = false;
      const unionErrors: ZodError[] = [];
      for (const option of def.options) {
        try {
          option.parse(obj);
```

This way validation isn't skipped for the `id` property for the second item in the union (from my example earlier). I have run the test suite and the test doesn't fail, however given it's recursive nature i don't feel confident that this really fixes the problem. Another possible solution would be to pass `parse` a modified `params` object: `{...params, seen: []}`.

**Comment by vriad** [comment]
Hey Jacco - 

Good issue. As you deduced, this is related to the `seen` param, which is the (quite dumb) method Zod uses to prevent infinite loops with recursive data. As you can see it has some intended consequences. I already implemented a fix that should solve this problem, but I'm still testing it, should get published sometime this weekend.👍

Colin

**Comment by vriad** [comment]
@flenter fixed in zod@1.11. Thanks for reporting this and providing such a detailed writeup!
